tools/tag.py

```diff
@@ -0,0 +1,34 @@
+#!/usr/bin/env python3
+
+import sys
+
+if len(sys.argv) < 2:
+    print("Provide version string")
+    sys.exit(1)
+
+orig = sys.argv[1]
+
+class Version:
+
+    def __init__(self, major, minor, patch, release, commit):
+        self.major = major
+        self.minor = minor
+        self.patch = patch
+        self.release = release
+        self.commit = commit
+
+    @staticmethod
+    def fromstring(s):
+        major, minor, s = s.split(".", 2)
+        patch, build, commit = s.split("-", 2)
+        return Version(major, minor, patch, build, commit)
+
+    def __str__(self):
+        return f"{self.major}.{self.minor}.{self.patch}-{self.release}-{self.commit}"
+
+
+if __name__ == "__main__":
+    v = Version.fromstring(orig)
+    v.minor = str(int(v.minor) + 1)
```

Comment:
This feels wrong, considering we still have the `a.b.x` GitHub tag, it will produce `a.(b+1).0-...`.

I am asking myself, can we bump the `a.b.x` to the next minor release?

https://github.com/stackrox/stackrox/blob/master/.github/workflows/start-release.yml#L157-L162

Reply:
I'm not sure I follow the concern.  Would this (`tag.py`) script be used to create the version string for build on the release branches as well?  Based on a quick look it seems like `BUILD_TAG` is set in downstream builds, thus overriding what this script would produce.

Reply:
I'm with Tom, the release automation scripting should place `A.(B+1).x` tag instead of `A.B.x` (which it currently does) on `master` when `release-A.B` branch is cut.

Reply:
> Would this (tag.py) script be used to create the version string for build on the release branches as well?

It would be used for all commits, except tagged ones. 

There is one slight caveat with doing the `A.(B+1).x` approach. We tag the commit where we branch off the release branch. Any commit on the release branch before the first release candidate on the `.0` release (which is the `A.B.0-rc.1` tag), would have `A.(B+1).x-n-g123456`. This might be ignored, because those commits are not built - they are updating the CHANGELOG, cherry-picks for the first RC and pushed with an empty commit. See https://github.com/stackrox/stackrox/commits/release-4.4/?before=53dd371292a8277e1049815c4f408d04a041f317+105 - 7fc738558fd7e77532816bbf0e5f79c1e7c79eec is pushed, preceding commits until 1f75c80c687851606cbfece6ccd46978d59293a4 are not built. 

A "solution" would be to add an empty commit with the tag to the master branch after the branching point - that leaves the question how to tag if the branching point is not the latest commit on master - the branching may be done couple of hours/days after the code freeze at a specific commit.

Reply:
Is there a reason not to tag the very first commit on the `release-A.B` branch with `A.B.0-rc.0`? Also, the very first commit could be an empty placeholder so that there's always _something_.

Reply:
> Is there a reason not to tag the very first commit on the release-A.B branch with A.B.0-rc.0? Also, the very first commit could be an empty placeholder so that there's always something.

Not really. Tagged commits have a few additional checks, like confirming that COLLECTOR_VERSION and SCANNER_VERSION are SemVer, but failure here may be informative, for the release engineers - if they look at the status of this tag at all.

Reply:
I think, we converged on the approach so I'm resolving this thread.